### PR TITLE
FreeBSD: fix panic due to tqid overflow

### PR DESCRIPTION
**Pull request by mjguzik**
The 32-bit counter eventually wraps to 0 which is a sentinel for invalid
id.

Make it 64-bit on LP64 platforms and 0-check otherwise.

Note: Linux counterpart uses id stored per queue instead of a global.
I did not check going that way is feasible with the goal being the
minimal fix doing the job.

Signed-off-by: Mateusz Guzik <mjguzik@gmail.com>

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/os/freebsd/spl/spl_taskq.c [hunk 1]
```diff
@@ -124,6 +123,28 @@ system_taskq_fini(void *arg)
 SYSUNINIT(system_taskq_fini, SI_SUB_CONFIGURE, SI_ORDER_ANY, system_taskq_fini,
     NULL);
 
+#ifdef __LP64__
+static taskqid_t
+__taskq_genid(void)
+{
+
+	return (atomic_fetchadd_long(&tqidnext, 1) + 1);
```
Comment (freqlabs) [inline module/os/freebsd/spl/spl_taskq.c:131]: Extra +1?

File: module/os/freebsd/spl/spl_taskq.c [hunk 2]
```diff
@@ -124,6 +123,28 @@ system_taskq_fini(void *arg)
 SYSUNINIT(system_taskq_fini, SI_SUB_CONFIGURE, SI_ORDER_ANY, system_taskq_fini,
     NULL);
 
+#ifdef __LP64__
+static taskqid_t
+__taskq_genid(void)
+{
+
```
Comment (freqlabs) [inline module/os/freebsd/spl/spl_taskq.c:130]: Nit: blank line not needed.

File: module/os/freebsd/spl/spl_taskq.c [hunk 3]
```diff
@@ -124,6 +123,28 @@ system_taskq_fini(void *arg)
...
+	return (atomic_fetchadd_long(&tqidnext, 1) + 1);
+}
+#else
+static taskqid_t
+__taskq_genid(void)
+{
+	taskqid_t tqid;
+
+	for (;;) {
+		tqid = atomic_fetchadd_int(&tqidnext, 1) + 1;
```
Comment (freqlabs) [inline module/os/freebsd/spl/spl_taskq.c:140]: Ditto extra +1

File: module/os/freebsd/spl/spl_taskq.c [hunk 4]
```diff
@@ -70,7 +70,7 @@ static unsigned long tqenthash;
 static unsigned long tqenthashlock;
 static struct sx *tqenthashtbl_lock;
 
-static uint32_t tqidnext = 1;
+static taskqid_t tqidnext;
```
Comment (freqlabs) [inline module/os/freebsd/spl/spl_taskq.c:73]: I prefer the explicit initialization for clarity, and starting at 1 avoids the extra +1.

File: module/os/freebsd/spl/spl_taskq.c (hunk 1 above)
Comment (mjguzik) [reply module/os/freebsd/spl/spl_taskq.c:131]: This is the idiomatic use, you can grep the tree to check. Semantics of the routine are "add 1, but return whatever was there prior to it", meaning patched code uses the very id it bumped to, while original code uses the previous one.

Comment (freqlabs) [reply module/os/freebsd/spl/spl_taskq.c:131]: In that case perhaps we should be using `atomic_add_64_nv` from `sys/atomic.h` (in spl).
```C
static inline uint64_t
atomic_add_64_nv(volatile uint64_t *target, int64_t delta)
{
        return (atomic_fetchadd_64(target, delta) + delta);
}
```

**Review by freqlabs — APPROVED** [review]
LGTM, no need to hold this up on a few nitpicks.

File: module/os/freebsd/spl/spl_taskq.c [hunk 5]
```diff
@@ -143,8 +164,10 @@ taskq_lookup(taskqid_t tqid)
 static taskqid_t
 taskq_insert(taskq_ent_t *ent)
 {
-	taskqid_t tqid = atomic_fetchadd_int(&tqidnext, 1);
+	taskqid_t tqid;
 
+	tqid = __taskq_genid();
+	VERIFY(tqid);
```
Comment (behlendorf) [inline module/os/freebsd/spl/spl_taskq.c:170]: An `ASSERT` at the end of `__taskq_genid()` seems sufficient for this and would cover all callers.

Comment (mjguzik) [reply module/os/freebsd/spl/spl_taskq.c]: Well there are two and merging them into one is imo uglier. Still, pushed asserts into them.

**Review by freqlabs — APPROVED** [review]

**Review by behlendorf — APPROVED** [review]

**Pull request merged**